### ISSUE TITLE
Uncaught exception in ext/websocket/01_websocket.js:507:13

### ISSUE THREAD

**Issue by quentinadam** [denoland/deno#16676]
[Using Deno 1.28.0]

I frequently have the following exception that kills my program:

```
error: Uncaught (in promise) Error: IO error: Broken pipe (os error 32): Broken pipe (os error 32)
at async deno:ext/websocket/01_websocket.js:507:13
```

I had a look at the source code, and it seems that this set of lines should be included in a try-catch block:

https://github.com/denoland/deno/blob/1416713cb3af8a952b1ae9952091706e2540341c/ext/websocket/01_websocket.js#L507

```
await core.opAsync("op_ws_send", this[_rid], {
  kind: "ping",
});
```

**Comment by quentinadam** [comment]
Hello,

I continue to have this bug which is quite annoying.

Here is the stack trace in In deno 1.29.1:

```
Error 'IO error: Broken pipe (os error 32)' contains boxed error of unknown type:
  Io(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
  Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }
error: Uncaught (in promise) Error: IO error: Broken pipe (os error 32): Broken pipe (os error 32)
    at async deno:ext/websocket/01_websocket.js:514:13
```

**Comment by bartlomieju** [comment]
Does this problem still happen on the latest version?

**Comment by quentinadam** [comment]
I would need to check, but it's slightly annoying because I can't upgrade past Deno 1.31 because otherwise I am getting memory leak errors every few hours. (https://github.com/denoland/deno/issues/18369).

**Comment by crowlKats** [comment]
@quentinadam does this still occur?

**Comment by bartlomieju** [comment]
@quentinadam You can upgrade to Deno v1.33 (or later for the time being) that works around the memory leak - we're still waiting on V8 11.7 release to properly fix the problem from that issue.

**Comment by milaz** [comment]
I have closed #17823, as it is most likely a duplicate of this issue.
However, there may be some useful additional info.

While everyone here is talking about a reproduction, I'd like to note that you can see a problem just by looking at the code.
ext/websocket/01_websocket.js line 505, `setTimeout`. What happens when there is an uncaught exception in a function called from `setTimout`? A crash. Line 507, `await core.opAsync("op_ws_send", ...)` a function that can raise an exception. And it does, at the most inconvenient moment.

I would be extremely grateful if you finally wrapped that function with try .. catch block, please.
I would have done it myself, but I still believe you have more knowledge to decide what amount of adjacent code should go into that block, what kind of exceptions does `opAsync` raise and what to do best when you are handling them, especially regarding the state of the socket.

There are two `setTimeout`s there with no exception handlers. It would be great to check those out in a single fix.

**Comment by littledivy** [comment]
Hi, This was fixed in the latest release (1.36.1). Feel free to reopen if its still happening.

**Comment by jflatow** [comment]
Has there been a regression? I'm using 1.37.1 and getting ```error: Uncaught (in promise) TypeError: Broken pipe (os error 32)
    at async ext:deno_websocket/01_websocket.js:517:15``` on a websocket over tls (but not without)
